backend/src/create/dog_leg.rs

```diff
@@ -0,0 +1,157 @@
+        }
+    }
+    info!("Collapsing {} dog-legs", dog_legs.len());
+
+    for (collapse_e, side1, side2) in dog_legs {
+        fix_dog_leg(graph, collapse_e, side1, side2);
+    }
+
+    // Remove deleted edges from node_to_edge, used later for applying existing modal filters.
+    graph
+        .node_to_edge
+        .retain(|_, e| graph.edges.contains_key(e));
+}
+
+/// If this edge is the short edge causing a dog-leg intersection, then return the two
+/// perpendicular "side roads".
+fn detect_dog_leg(graph: &Graph, e: EdgeID) -> Option<(EdgeID, EdgeID)> {
+    let edge = &graph.edges[&e];
+    if Euclidean.length(&edge.linestring) > 5.0 {
+        return None;
+    }
+    if !edge.osm_tags.has("name") {
+        return None;
+    }
+    let src_i = &graph.intersections[&edge.src];
+    let dst_i = &graph.intersections[&edge.dst];
+    let mut src_edges = src_i.edges.clone();
+    let mut dst_edges = dst_i.edges.clone();
+    if src_edges.len() != 3 || dst_edges.len() != 3 {
+        return None;
+    }
+
+    // Find the two "side roads" with a different name than the short edge
+    // (TODO We could use angle to be safer than name, but name works fine in all known cases so
+    // far)
+    src_edges.retain(|x| graph.edges[x].osm_tags.get("name") != edge.osm_tags.get("name"));
+    dst_edges.retain(|x| graph.edges[x].osm_tags.get("name") != edge.osm_tags.get("name"));
+
+    if src_edges.len() != 1 || dst_edges.len() != 1 {
+        return None;
+    }
+
+    let src_edge = &graph.edges[&src_edges[0]];
+    let dst_edge = &graph.edges[&dst_edges[0]];
+
+    // Check the bearing of the two side roads, both pointing "away" from the short edge
+    let b1 = bearing_from_endpoint(src_i.point, &src_edge.linestring);
+    let b2 = bearing_from_endpoint(dst_i.point, &dst_edge.linestring);
+
+    // If these both point about the same way, they're on the "same side" of the short edge.
+    // Not a dog leg.
+    if (b1 - b2).abs() < 30.0 {
+        return None;
+    }
+
+    Some((src_edges[0], dst_edges[0]))
+}
+
+fn fix_dog_leg(graph: &mut Graph, collapse_e: EdgeID, side1: EdgeID, side2: EdgeID) {
+    let (src, dst, midpt) = {
+        let edge = &graph.edges[&collapse_e];
+        let midpt = edge
+            .linestring
+            .point_at_ratio_from_start(&Euclidean, 0.5)
+            .unwrap();
+        (edge.src, edge.dst, midpt)
+    };
+
+    // Create a new intersection at the middle of the short edge
+    let new_intersection = new_intersection_id(graph);
+    graph.intersections.insert(
+        new_intersection,
+        Intersection {
+            id: new_intersection,
+            edges: Vec::new(),
+            point: midpt,
+            // OSM IDs are only used for debugging. Rather than change the representation here to
+            // allow for "synthetic" intersections without an OSM node, just borrow one of the
+            // endpoints.
+            osm_node: graph.intersections[&src].osm_node,
```

Comment:
I understand how this currently isn't causing any problems, but it might be confusing at some point. One thing I sometimes see with synthetic features is using a negative id as a sentinel value. It doesn't seem like a big deal though.